### PR TITLE
[lld][ELF] Support LLVM repository and LLVM revision information

### PR DESCRIPTION
**Pull request by hongyu-dev**
Added LLVM repository and LLVM revision information for `lld::getLLDVersion()`

Before this change:
```
hongyuchy@hongyuchy:~/llvm-project/.build_lld_version$ bin/ld.lld --version
LLD 19.0.0 (compatible with GNU linkers) 
```

After this change with LLVM_APPEND_VC_REV=on
```
hongyuchy@hongyuchy:~/llvm-project/.build_lld_version$ bin/ld.lld --version
LLD 19.0.0 (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309), compatible with GNU linkers
```

with LLVM_APPEND_VC_REV=off
```
hongyuchy@hongyuchy:~/llvm-project/.build_lld_version$ bin/ld.lld --version
LLD 19.0.0, compatible with GNU linkers
```

### CODE REVIEW

File: lld/ELF/Driver.cpp [hunk 1]
```diff
@@ -164,6 +164,7 @@ bool link(ArrayRef<const char *> args, llvm::raw_ostream &stdoutOS,
 
   return errorCount() == 0;
 }
+
```
Comment (Prabhuk) [inline lld/ELF/Driver.cpp:167]: Remove empty line.

**Review by sbc100 — COMMENTED** [review]
Could you also update wasm-ld?

**Comment by petrhosek** [comment]
I think this change should instead be made to https://github.com/llvm/llvm-project/blob/f9efc295089217425d8220af892dcc5ede9eb9f7/lld/Common/Version.cpp so it applies to all LLD drivers uniformly.

Specifically, the comment above `getLLDVersion` says:
https://github.com/llvm/llvm-project/blob/f9efc295089217425d8220af892dcc5ede9eb9f7/lld/Common/Version.cpp#L17-L19

That's not the case though since the string doesn't include the repository or the revision information, so I'd change the implementation to match the comment.

I'd also consider removing the parenthesis around the `compatible with GNU linkers` part in the ELF driver so we end up with:

```
LLD 19.0.0 (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309) compatible with GNU linkers
```

I think this is a bit cleaner than:

```
LLD 19.0.0 (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309) (compatible with GNU linkers)
```

**Comment by compnerd** [comment]
I think that we should check autotools before changing the "(compatible with GNU linkers)" as it may have the parenthesis hardcoded.

File: lld/ELF/Driver.cpp (hunk 1 above)
Comment (yugier) [reply lld/ELF/Driver.cpp]: I just did!

**Comment by brad0** [comment]
> I think that we should check autotools before changing the "(compatible with GNU linkers)" as it may have the parenthesis hardcoded.

The string itself does not matter. It is just checking for the name GNU.

```
      case `"$lt_cv_path_LD" -v 2>&1 </dev/null` in
      *GNU* | *'with BFD'*)
```

File: lld/Common/Version.cpp [hunk 2]
```diff
@@ -22,6 +22,10 @@ std::string lld::getLLDVersion() {
 #define LLD_VENDOR_DISPLAY LLD_VENDOR " "
 #else
 #define LLD_VENDOR_DISPLAY
+#endif
+#if defined(LLVM_REPOSITORY) || defined(LLVM_REVISION)
+  return LLD_VENDOR_DISPLAY "LLD " LLD_VERSION_STRING " (" LLVM_REPOSITORY
+                            " " LLVM_REVISION ")";
 #endif
   return LLD_VENDOR_DISPLAY "LLD " LLD_VERSION_STRING;
```
Comment (petrhosek) [inline lld/Common/Version.cpp:30]:
```suggestion
#if defined(LLVM_REPOSITORY) || defined(LLVM_REVISION)
  return LLD_VENDOR_DISPLAY "LLD " LLD_VERSION_STRING " (" LLVM_REPOSITORY
                            " " LLVM_REVISION ")";
#else
  return LLD_VENDOR_DISPLAY "LLD " LLD_VERSION_STRING;
#endif
```

**Review by petrhosek — APPROVED** [review]
Please wait for @MaskRay to take a look as well.

File: lld/Common/Version.cpp (hunk 2 above)
Comment (yugier) [reply lld/Common/Version.cpp]: updated!

**Comment by MaskRay** [comment]
The description (which will be used as the default commit message when "Squash and merge" is clicked) is probably about 
`LLVM_APPEND_VC_REV=on`. 

> I'd also consider removing the parenthesis around the compatible with GNU linkers part in the ELF driver so we end up with:

In the `LLVM_APPEND_VC_REV=off` configuration,

```
% fld.lld --version
LLD 19.0.0 compatible with GNU linkers
```

Without parentheses, "compatible ..." might be parsed as part of the version string.

Should parentheses be added back? We could also use a leading comma or something else, but lld/MinGW/Driver.cpp should be updated as well for consistency.

**Comment by petrhosek** [comment]
> The description (which will be used as the default commit message when "Squash and merge" is clicked) is probably about `LLVM_APPEND_VC_REV=on`.
> 
> > I'd also consider removing the parenthesis around the compatible with GNU linkers part in the ELF driver so we end up with:
> 
> In the `LLVM_APPEND_VC_REV=off` configuration,
> 
> ```
> % fld.lld --version
> LLD 19.0.0 compatible with GNU linkers
> ```
> 
> Without parentheses, "compatible ..." might be parsed as part of the version string.
> 
> Should parentheses be added back? We could also use a leading comma or something else, but lld/MinGW/Driver.cpp should be updated as well for consistency.

Looks like [mold uses the following format](https://github.com/rui314/mold/blob/b145377c39de8ed9b6ee3b1106cdf206fafe1605/elf/config.cc): 

```
$ ld.lld --version
LLD 19.0.0 (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309; compatible with GNU ld)
```

Should we use the same format for consistency?

**Comment by MaskRay** [comment]
We could use a comma to be consistent with other LLVM binary utilities: `llvm-objcopy, compatible with GNU objcopy`/
`llvm-nm, compatible with GNU nm`.

**Comment by yugier** [comment]
> We could use a comma to be consistent with other LLVM binary utilities: `llvm-objcopy, compatible with GNU objcopy`/ `llvm-nm, compatible with GNU nm`.

Do you prefer we have 
```
$ ld.lld --version
LLD 19.0.0 (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309), compatible with GNU ld
```

or 
```
$ ld.lld --version
LLD 19.0.0, compatible with GNU ld (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309)
```

Thank you!

**Comment by MaskRay** [comment]
`LLD 19.0.0 (https://github.com/yugier/llvm-project.git 4134b33c6a362cb462b335177d6d9e8235f04309), compatible with GNU ld` to keep the version parts together.

**Review by MaskRay — APPROVED** [review]
LGTM. The `After this change:` part in the description should be adjusted.
The output is probably for LLVM_APPEND_VC_REV=on

**Comment by yugier** [comment]
> LGTM. The `After this change:` part in the description should be adjusted. The output is probably for LLVM_APPEND_VC_REV=on

I just did! Thank you for your fast response and help! :)

**Comment by mstorsjo** [comment]
Sorry, but this change has broken how Meson detects and disambiguates between the `ld.lld` and `lld-link` interfaces for Windows targets. See https://github.com/mesonbuild/meson/blob/1.4.1/mesonbuild/linkers/detect.py#L67 - Meson looks for literally the string `(compatible with GNU linkers)` including the parentheses.

I would request reinstating that form.

We should of course request Meson to match a less strict pattern here as well, but even then, we should work with the variety of existing Meson versions out there as well, at least for a couple years after that has been changed.

**Comment by mstorsjo** [comment]
As this check in Meson only seems to be used to differentiate between the `ld.lld` and `lld-link` interface for the Windows targets, it should be enough to only adjust the printout in `lld/MinGW/Driver.cpp`, e.g. with this diff:

```diff
diff --git a/lld/MinGW/Driver.cpp b/lld/MinGW/Driver.cpp
index 1fd120ad3601..35fd478a2190 100644
--- a/lld/MinGW/Driver.cpp
+++ b/lld/MinGW/Driver.cpp
@@ -199,7 +199,7 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
   // a GNU compatible linker. As long as an output for the -v option
   // contains "GNU" or "with BFD", they recognize us as GNU-compatible.
   if (args.hasArg(OPT_v) || args.hasArg(OPT_version))
-    message(getLLDVersion() + ", compatible with GNU linkers");
+    message(getLLDVersion() + " (compatible with GNU linkers)");
 
   // The behavior of -v or --version is a bit strange, but this is
   // needed for compatibility with GNU linkers.

```

**Comment by mstorsjo** [comment]
I posted such a change in https://github.com/llvm/llvm-project/pull/97698, and made a PR to Meson to loosen the check now in https://github.com/mesonbuild/meson/pull/13383.

**Comment by yamader** [comment]
This change breaks scripts that want to detect the LLD version when `LLLVM_APPEND_VC_REV=off`.
The second block of `ld.lld --version` should only contain the version number.

For example, this change broke a very basic Linux config script.
It expects the second block of `--version` to contain only a period-separated numeric version string.
https://github.com/torvalds/linux/blob/e24b3ff/scripts/ld-version.sh#L50

**Pull request merged**